File: examples/java/DijkstraExamples.java
import java.util.*;

public class DijkstraExamples {
    
    // 1. Network Delay Time
    public int networkDelayTime(int[][] times, int n, int k) {
        Map<Integer, List<int[]>> graph = new HashMap<>();
        for (int[] time : times) {
            graph.computeIfAbsent(time[0], x -> new ArrayList<>()).add(new int[]{time[1], time[2]});
        }
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        pq.offer(new int[]{k, 0});
        
        Set<Integer> visited = new HashSet<>();
        int maxTime = 0;
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int node = curr[0], time = curr[1];
            
            if (visited.contains(node)) continue;
            visited.add(node);
            maxTime = Math.max(maxTime, time);
            
            if (graph.containsKey(node)) {
                for (int[] edge : graph.get(node)) {
                    int next = edge[0], weight = edge[1];
                    if (!visited.contains(next)) {
                        pq.offer(new int[]{next, time + weight});
                    }
                }
            }
        }
        
        return visited.size() == n ? maxTime : -1;
    }
    
    // 2. Cheapest Flights Within K Stops
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        Map<Integer, List<int[]>> graph = new HashMap<>();
        for (int[] flight : flights) {
            graph.computeIfAbsent(flight[0], x -> new ArrayList<>()).add(new int[]{flight[1], flight[2]});
        }
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        pq.offer(new int[]{src, 0, 0});
        
        Map<Integer, Integer> visited = new HashMap<>();
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int node = curr[0], cost = curr[1], stops = curr[2];
            
            if (node == dst) return cost;
            if (stops > k) continue;
            if (visited.containsKey(node) && visited.get(node) <= stops) continue;
            
            visited.put(node, stops);
            
            if (graph.containsKey(node)) {
                for (int[] edge : graph.get(node)) {
                    int next = edge[0], price = edge[1];
                    pq.offer(new int[]{next, cost + price, stops + 1});
                }
            }
        }
        
        return -1;
    }
    
    // 3. Path with Maximum Probability
    public double maxProbability(int n, int[][] edges, double[] succProb, int start, int end) {
        Map<Integer, List<double[]>> graph = new HashMap<>();
        for (int i = 0; i < edges.length; i++) {
            int u = edges[i][0], v = edges[i][1];
            double prob = succProb[i];
            graph.computeIfAbsent(u, x -> new ArrayList<>()).add(new double[]{v, prob});
            graph.computeIfAbsent(v, x -> new ArrayList<>()).add(new double[]{u, prob});
        }
        
        PriorityQueue<double[]> pq = new PriorityQueue<>((a, b) -> Double.compare(b[1], a[1]));
        pq.offer(new double[]{start, 1.0});
        
        Set<Integer> visited = new HashSet<>();
        
        while (!pq.isEmpty()) {
            double[] curr = pq.poll();
            int node = (int)curr[0];
            double prob = curr[1];
            
            if (node == end) return prob;
            if (visited.contains(node)) continue;
            visited.add(node);
            
            if (graph.containsKey(node)) {
                for (double[] edge : graph.get(node)) {
                    int next = (int)edge[0];
                    double nextProb = edge[1];
                    if (!visited.contains(next)) {
                        pq.offer(new double[]{next, prob * nextProb});
                    }
                }
            }
        }
        
        return 0.0;
    }
    
    // 4. Path With Minimum Effort
    public int minimumEffortPath(int[][] heights) {
        int m = heights.length, n = heights[0].length;
        int[][] dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        pq.offer(new int[]{0, 0, 0});
        
        boolean[][] visited = new boolean[m][n];
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int x = curr[0], y = curr[1], effort = curr[2];
            
            if (x == m - 1 && y == n - 1) return effort;
            if (visited[x][y]) continue;
            visited[x][y] = true;
            
            for (int[] dir : dirs) {
                int nx = x + dir[0], ny = y + dir[1];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n && !visited[nx][ny]) {
                    int newEffort = Math.max(effort, Math.abs(heights[nx][ny] - heights[x][y]));
                    pq.offer(new int[]{nx, ny, newEffort});
                }
            }
        }
        
        return -1;
    }
    
    // 5. Find the City With the Smallest Number of Neighbors at a Threshold Distance
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        int[][] dist = new int[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(dist[i], Integer.MAX_VALUE);
            dist[i][i] = 0;
        }
        
        for (int[] edge : edges) {
            dist[edge[0]][edge[1]] = edge[2];
            dist[edge[1]][edge[0]] = edge[2];
        }
        
        // Floyd-Warshall
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    if (dist[i][k] != Integer.MAX_VALUE && dist[k][j] != Integer.MAX_VALUE) {
                        dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                    }
                }
            }
        }
        
        int minCount = n, result = 0;
        for (int i = 0; i < n; i++) {
            int count = 0;
            for (int j = 0; j < n; j++) {
                if (i != j && dist[i][j] <= distanceThreshold) count++;
            }
            if (count <= minCount) {
                minCount = count;
                result = i;
            }
        }
        
        return result;
    }
    
    // 6. Minimum Cost to Make at Least One Valid Path in a Grid
    public int minCost(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        int[][] dirs = {{0,1}, {0,-1}, {1,0}, {-1,0}};
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        pq.offer(new int[]{0, 0, 0});
        
        boolean[][] visited = new boolean[m][n];
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int x = curr[0], y = curr[1], cost = curr[2];
            
            if (x == m - 1 && y == n - 1) return cost;
            if (visited[x][y]) continue;
            visited[x][y] = true;
            
            for (int dir = 0; dir < 4; dir++) {
                int nx = x + dirs[dir][0], ny = y + dirs[dir][1];
                if (nx >= 0 && nx < m && ny >= 0 && ny < n && !visited[nx][ny]) {
                    int newCost = cost + (grid[x][y] == dir + 1 ? 0 : 1);
                    pq.offer(new int[]{nx, ny, newCost});
                }
            }
        }
        
        return -1;
    }
    
    // 7. Number of Ways to Arrive at Destination
    public int countPaths(int n, int[][] roads) {
        int MOD = 1000000007;
        Map<Integer, List<long[]>> graph = new HashMap<>();
        
        for (int[] road : roads) {
            graph.computeIfAbsent(road[0], x -> new ArrayList<>()).add(new long[]{road[1], road[2]});
            graph.computeIfAbsent(road[1], x -> new ArrayList<>()).add(new long[]{road[0], road[2]});
        }
        
        PriorityQueue<long[]> pq = new PriorityQueue<>((a, b) -> Long.compare(a[1], b[1]));
        pq.offer(new long[]{0, 0});
        
        long[] dist = new long[n];
        long[] ways = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[0] = 0;
        ways[0] = 1;
        
        while (!pq.isEmpty()) {
            long[] curr = pq.poll();
            int node = (int)curr[0];
            long time = curr[1];
            
            if (time > dist[node]) continue;
            
            if (graph.containsKey(node)) {
                for (long[] edge : graph.get(node)) {
                    int next = (int)edge[0];
                    long weight = edge[1];
                    long newTime = time + weight;
                    
                    if (newTime < dist[next]) {
                        dist[next] = newTime;
                        ways[next] = ways[node];
                        pq.offer(new long[]{next, newTime});
                    } else if (newTime == dist[next]) {
                        ways[next] = (ways[next] + ways[node]) % MOD;
                    }
                }
            }
        }
        
        return (int)ways[n - 1];
    }
    
    // 8. Swim in Rising Water
    public int swimInWater(int[][] grid) {
        int n = grid.length;
        int[][] dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        pq.offer(new int[]{0, 0, grid[0][0]});
        
        boolean[][] visited = new boolean[n][n];
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int x = curr[0], y = curr[1], time = curr[2];
            
            if (x == n - 1 && y == n - 1) return time;
            if (visited[x][y]) continue;
            visited[x][y] = true;
            
            for (int[] dir : dirs) {
                int nx = x + dir[0], ny = y + dir[1];
                if (nx >= 0 && nx < n && ny >= 0 && ny < n && !visited[nx][ny]) {
                    pq.offer(new int[]{nx, ny, Math.max(time, grid[nx][ny])});
                }
            }
        }
        
        return -1;
    }
    
    // 9. Shortest Path in Binary Matrix
    public int shortestPathBinaryMatrix(int[][] grid) {
        int n = grid.length;
        if (grid[0][0] == 1 || grid[n-1][n-1] == 1) return -1;
        
        int[][] dirs = {{-1,-1}, {-1,0}, {-1,1}, {0,-1}, {0,1}, {1,-1}, {1,0}, {1,1}};
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[2] - b[2]);
        pq.offer(new int[]{0, 0, 1});
        
        boolean[][] visited = new boolean[n][n];
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int x = curr[0], y = curr[1], dist = curr[2];
            
            if (x == n - 1 && y == n - 1) return dist;
            if (visited[x][y]) continue;
            visited[x][y] = true;
            
            for (int[] dir : dirs) {
                int nx = x + dir[0], ny = y + dir[1];
                if (nx >= 0 && nx < n && ny >= 0 && ny < n && 
                    grid[nx][ny] == 0 && !visited[nx][ny]) {
                    pq.offer(new int[]{nx, ny, dist + 1});
                }
            }
        }
        
        return -1;
    }
    
    // 10. Minimum Spanning Tree (Kruskal's Algorithm)
    public int minimumSpanningTree(int n, int[][] edges) {
        Arrays.sort(edges, (a, b) -> a[2] - b[2]);
        UnionFind uf = new UnionFind(n);
        int cost = 0, edgesUsed = 0;
        
        for (int[] edge : edges) {
            if (uf.union(edge[0], edge[1])) {
                cost += edge[2];
                edgesUsed++;
                if (edgesUsed == n - 1) break;
            }
        }
        
        return edgesUsed == n - 1 ? cost : -1;
    }
    
    // Union-Find for MST
    class UnionFind {
        int[] parent, rank;
        
        public UnionFind(int n) {
            parent = new int[n];
            rank = new int[n];
            for (int i = 0; i < n; i++) parent[i] = i;
        }
        
        public int find(int x) {
            if (parent[x] != x) parent[x] = find(parent[x]);
            return parent[x];
        }
        
        public boolean union(int x, int y) {
            int px = find(x), py = find(y);
            if (px == py) return false;
            if (rank[px] < rank[py]) {
                parent[px] = py;
            } else if (rank[px] > rank[py]) {
                parent[py] = px;
            } else {
                parent[py] = px;
                rank[px]++;
            }
            return true;
        }
    }
    
    public static void main(String[] args) {
        DijkstraExamples examples = new DijkstraExamples();
        
        System.out.println("=== Dijkstra's Algorithm Examples ===");
        
        // Test 1: Network Delay Time
        int[][] times = {{2,1,1},{2,3,1},{3,4,1}};
        System.out.println("1. Network Delay Time: " + examples.networkDelayTime(times, 4, 2));
        
        // Test 2: Cheapest Flights
        int[][] flights = {{0,1,100},{1,2,100},{0,2,500}};
        System.out.println("2. Cheapest Flights: " + examples.findCheapestPrice(3, flights, 0, 2, 1));
        
        // Test 3: Max Probability
        int[][] edges = {{0,1},{1,2},{0,2}};
        double[] succProb = {0.5,0.5,0.2};
        System.out.println("3. Max Probability: " + examples.maxProbability(3, edges, succProb, 0, 2));
        
        // Test 4: Minimum Effort Path
        int[][] heights = {{1,2,2},{3,8,2},{5,3,5}};
        System.out.println("4. Minimum Effort Path: " + examples.minimumEffortPath(heights));
        
        // Test 5: Find the City
        int[][] cityEdges = {{0,1,3},{1,2,1},{1,3,4},{2,3,1}};
        System.out.println("5. Find the City: " + examples.findTheCity(4, cityEdges, 4));
        
        // Test 6: Min Cost Valid Path
        int[][] grid = {{1,1,1,1},{2,2,2,2},{1,1,1,1},{2,2,2,2}};
        System.out.println("6. Min Cost Valid Path: " + examples.minCost(grid));
        
        // Test 7: Count Paths
        int[][] roads = {{0,6,7},{0,1,2},{1,2,3},{1,3,3},{6,3,3},{3,5,1},{6,5,1},{2,5,1},{0,4,5},{4,6,2}};
        System.out.println("7. Count Paths: " + examples.countPaths(7, roads));
        
        // Test 8: Swim in Water
        int[][] swimGrid = {{0,2},{1,3}};
        System.out.println("8. Swim in Water: " + examples.swimInWater(swimGrid));
        
        // Test 9: Shortest Path Binary Matrix
        int[][] binaryGrid = {{0,0,0},{1,1,0},{1,1,0}};
        System.out.println("9. Shortest Path Binary Matrix: " + examples.shortestPathBinaryMatrix(binaryGrid));
        
        // Test 10: Minimum Spanning Tree
        int[][] mstEdges = {{0,1,10},{0,2,6},{0,3,5},{1,3,15},{2,3,4}};
        System.out.println("10. Minimum Spanning Tree: " + examples.minimumSpanningTree(4, mstEdges));
    }
}

File: examples/python/dijkstra_examples.py
import heapq
from collections import defaultdict

# 1. Dijkstra
def dijkstra(graph, start):
    distances = {node: float('inf') for node in graph}
    distances[start] = 0
    pq = [(0, start)]
    
    while pq:
        current_dist, current = heapq.heappop(pq)
        
        if current_dist > distances[current]:
            continue
        
        for neighbor, weight in graph[current]:
            distance = current_dist + weight
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(pq, (distance, neighbor))
    
    return distances

# 2. Network Delay Time
def network_delay_time(times, n, k):
    graph = defaultdict(list)
    for u, v, w in times:
        graph[u].append((v, w))
    
    # Ensure all nodes exist in graph
    for i in range(1, n + 1):
        if i not in graph:
            graph[i] = []
    
    distances = dijkstra(graph, k)
    max_time = max(distances[i] for i in range(1, n + 1))
    return max_time if max_time != float('inf') else -1

# 3. Cheapest Flights Within K Stops
def cheapest_flights_within_k_stops(n, flights, src, dst, k):
    graph = defaultdict(list)
    for u, v, w in flights:
        graph[u].append((v, w))
    
    # (cost, node, stops)
    pq = [(0, src, 0)]
    visited = {}
    
    while pq:
        cost, node, stops = heapq.heappop(pq)
        
        if node == dst:
            return cost
        
        if stops > k or (node in visited and visited[node] <= stops):
            continue
        
        visited[node] = stops
        
        for neighbor, price in graph[node]:
            heapq.heappush(pq, (cost + price, neighbor, stops + 1))
    
    return -1

# 4. Minimum Effort Path
def minimum_effort_path(heights):
    m, n = len(heights), len(heights[0])
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    
    # (effort, row, col)
    pq = [(0, 0, 0)]
    efforts = [[float('inf')] * n for _ in range(m)]
    efforts[0][0] = 0
    
    while pq:
        effort, row, col = heapq.heappop(pq)
        
        if row == m - 1 and col == n - 1:
            return effort
        
        if effort > efforts[row][col]:
            continue
        
        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            if 0 <= nr < m and 0 <= nc < n:
                new_effort = max(effort, abs(heights[nr][nc] - heights[row][col]))
                if new_effort < efforts[nr][nc]:
                    efforts[nr][nc] = new_effort
                    heapq.heappush(pq, (new_effort, nr, nc))
    
    return 0

# 5. Swim In Water
def swim_in_water(grid):
    n = len(grid)
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    
    # (time, row, col)
    pq = [(grid[0][0], 0, 0)]
    visited = set()
    
    while pq:
        time, row, col = heapq.heappop(pq)
        
        if (row, col) in visited:
            continue
        
        visited.add((row, col))
        
        if row == n - 1 and col == n - 1:
            return time
        
        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            if 0 <= nr < n and 0 <= nc < n and (nr, nc) not in visited:
                heapq.heappush(pq, (max(time, grid[nr][nc]), nr, nc))
    
    return -1

# 6. Shortest Path Binary Matrix
def shortest_path_binary_matrix(grid):
    n = len(grid)
    if grid[0][0] == 1 or grid[n-1][n-1] == 1:
        return -1
    
    directions = [(-1,-1), (-1,0), (-1,1), (0,-1), (0,1), (1,-1), (1,0), (1,1)]
    
    # (distance, row, col)
    pq = [(1, 0, 0)]
    visited = set()
    
    while pq:
        dist, row, col = heapq.heappop(pq)
        
        if (row, col) in visited:
            continue
        
        visited.add((row, col))
        
        if row == n - 1 and col == n - 1:
            return dist
        
        for dr, dc in directions:
            nr, nc = row + dr, col + dc
            if (0 <= nr < n and 0 <= nc < n and 
                grid[nr][nc] == 0 and (nr, nc) not in visited):
                heapq.heappush(pq, (dist + 1, nr, nc))
    
    return -1

# 7. Find Cheapest Price
def find_cheapest_price(n, flights, src, dst, k):
    graph = defaultdict(list)
    for u, v, w in flights:
        graph[u].append((v, w))
    
    # (cost, node, stops)
    pq = [(0, src, 0)]
    costs = {}
    
    while pq:
        cost, node, stops = heapq.heappop(pq)
        
        if node == dst:
            return cost
        
        if stops > k:
            continue
        
        if (node, stops) in costs and costs[(node, stops)] <= cost:
            continue
        
        costs[(node, stops)] = cost
        
        for neighbor, price in graph[node]:
            heapq.heappush(pq, (cost + price, neighbor, stops + 1))
    
    return -1

# 8. Minimum Cost To Make Valid Path
def minimum_cost_to_make_valid_path(grid):
    m, n = len(grid), len(grid[0])
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    
    # (cost, row, col)
    pq = [(0, 0, 0)]
    costs = [[float('inf')] * n for _ in range(m)]
    costs[0][0] = 0
    
    while pq:
        cost, row, col = heapq.heappop(pq)
        
        if row == m - 1 and col == n - 1:
            return cost
        
        if cost > costs[row][col]:
            continue
        
        for i, (dr, dc) in enumerate(directions):
            nr, nc = row + dr, col + dc
            if 0 <= nr < m and 0 <= nc < n:
                new_cost = cost + (0 if grid[row][col] == i + 1 else 1)
                if new_cost < costs[nr][nc]:
                    costs[nr][nc] = new_cost
                    heapq.heappush(pq, (new_cost, nr, nc))
    
    return -1

# 9. Shortest Distance From All Buildings
def shortest_distance_from_all_buildings(grid):
    m, n = len(grid), len(grid[0])
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    
    buildings = []
    for i in range(m):
        for j in range(n):
            if grid[i][j] == 1:
                buildings.append((i, j))
    
    total_distances = [[0] * n for _ in range(m)]
    reachable_count = [[0] * n for _ in range(m)]
    
    for building_idx, (start_r, start_c) in enumerate(buildings):
        # BFS from each building
        queue = [(start_r, start_c, 0)]
        visited = set()
        
        while queue:
            row, col, dist = queue.pop(0)
            
            if (row, col) in visited:
                continue
            
            visited.add((row, col))
            
            if grid[row][col] == 0:
                total_distances[row][col] += dist
                reachable_count[row][col] += 1
            
            for dr, dc in directions:
                nr, nc = row + dr, col + dc
                if (0 <= nr < m and 0 <= nc < n and 
                    (nr, nc) not in visited and grid[nr][nc] != 2):
                    queue.append((nr, nc, dist + 1))
    
    min_distance = float('inf')
    for i in range(m):
        for j in range(n):
            if (grid[i][j] == 0 and 
                reachable_count[i][j] == len(buildings)):
                min_distance = min(min_distance, total_distances[i][j])
    
    return min_distance if min_distance != float('inf') else -1

# 10. Minimum Spanning Tree Kruskal
def minimum_spanning_tree_kruskal(n, edges):
    # Union-Find for Kruskal's algorithm
    parent = list(range(n))
    
    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]
    
    def union(x, y):
        px, py = find(x), find(y)
        if px != py:
            parent[px] = py
            return True
        return False
    
    edges.sort(key=lambda x: x[2])
    mst_cost = 0
    edges_used = 0
    
    for u, v, weight in edges:
        if union(u, v):
            mst_cost += weight
            edges_used += 1
            if edges_used == n - 1:
                break
    
    return mst_cost if edges_used == n - 1 else -1

# 11. Run Examples
def run_examples():
    print("=== Dijkstra's Algorithm Examples ===")
    
    # Test 1: Basic Dijkstra
    graph = {
        'A': [('B', 4), ('C', 2)],
        'B': [('C', 1), ('D', 5)],
        'C': [('D', 8), ('E', 10)],
        'D': [('E', 2)],
        'E': []
    }
    print(f"Dijkstra from A: {dijkstra(graph, 'A')}")
    
    # Test 2: Network Delay Time
    times = [[2,1,1],[2,3,1],[3,4,1]]
    n, k = 4, 2
    print(f"Network Delay Time: {network_delay_time(times, n, k)}")
    
    # Test 3: Cheapest Flights
    n = 3
    flights = [[0,1,100],[1,2,100],[0,2,500]]
    src, dst, k = 0, 2, 1
    print(f"Cheapest Flights: {cheapest_flights_within_k_stops(n, flights, src, dst, k)}")
    
    # Test 4: Path With Minimum Effort
    heights = [[1,2,2],[3,8,2],[5,3,5]]
    print(f"Minimum Effort Path: {minimum_effort_path(heights)}")
    
    # Test 5: Swim in Rising Water
    grid = [[0,2],[1,3]]
    print(f"Swim in Water: {swim_in_water(grid)}")
    
    # Test 6: Shortest Path in Binary Matrix
    grid = [[0,0,0],[1,1,0],[1,1,0]]
    print(f"Shortest Path Binary Matrix: {shortest_path_binary_matrix(grid)}")
    
    # Test 7: Find Cheapest Price (duplicate function)
    n = 4
    flights = [[0,1,1],[0,2,5],[1,2,1],[2,3,1]]
    src, dst, k = 0, 3, 1
    print(f"Find Cheapest Price: {find_cheapest_price(n, flights, src, dst, k)}")
    
    # Test 8: Minimum Cost Valid Path
    grid = [[1,1,1,1],[2,2,2,2],[1,1,1,1],[2,2,2,2]]
    print(f"Min Cost Valid Path: {minimum_cost_to_make_valid_path(grid)}")
    
    # Test 9: Shortest Distance Buildings
    grid = [[1,0,2,0,1],[0,0,0,0,0],[0,0,1,0,0]]
    print(f"Shortest Distance Buildings: {shortest_distance_from_all_buildings(grid)}")
    
    # Test 10: MST Kruskal
    n = 4
    edges = [[0,1,10],[0,2,6],[0,3,5],[1,3,15],[2,3,4]]
    print(f"Minimum Spanning Tree: {minimum_spanning_tree_kruskal(n, edges)}")

if __name__ == "__main__":
    run_examples()